packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/internal/KtorNetworkTransport.kt

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2021 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm.mongodb.internal
+
+import io.ktor.client.*
+import io.ktor.client.call.*
+import io.ktor.client.engine.cio.*
+import io.ktor.client.features.*
+import io.ktor.client.features.logging.*
+import io.ktor.client.request.*
+import io.ktor.client.statement.*
+import io.ktor.http.*
+import io.realm.interop.NetworkTransport
+import io.realm.interop.Response
+import io.realm.internal.platform.freeze
+import io.realm.internal.platform.runBlocking
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.CoroutineDispatcher
+
+class KtorNetworkTransport(
+    override val authorizationHeaderName: String = "Authorization",
+    override val customHeaders: Map<String, String> = mapOf(),
+    private val timeoutMs: Long,
+    private val dispatcher: CoroutineDispatcher,
+) : NetworkTransport {
+
+    private val client: HttpClient = getClient()
+
+    override fun sendRequest(
+        method: String,
+        url: String,
+        headers: Map<String, String>,
+        body: String,
+        usesRefreshToken: Boolean
+    ): Response {
+        try {
+            return runBlocking(dispatcher) {
+                val requestBuilderBlock: HttpRequestBuilder.() -> Unit = {
+                    headers {
+                        // 1. First of all add all custom headers
+                        customHeaders.forEach {
+                            append(it.key, it.value)
+                        }
+
+                        // 2. Then replace default authorization header with custom one if present
+                        headers
+//                            .filter {
+//                                // Ignore Content-Type header as it's handled by the engine
+//                                it.key != HttpHeaders.ContentType
+//                            }
+                            .toMutableMap().also { receivedHeaders ->
+                                val authorizationHeaderValue =
```

Comment:
Maybe refactor into its own method. Something like `replaceAuthorizationHeaderIfNeeded` or similar. At least I had a bit of hard time following the logic here.

Reply:
I just simplified the implementation so that you should now be able to reason about it